### PR TITLE
fix(template-compiler): refactor logic to simplifiy attribute name handling

### PR DESCRIPTION
**Pull request by ravijayaramappa**
## Details
Refactoring a code path that had some redundant logic. No specific reason to fix this now. Noticed the room for improvement and fix it.

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

### CODE REVIEW

**Review by pmdartus — APPROVED** [review]

**Review by ekashida — COMMENTED** [review]
Should we add a regression test for this? It's not immediately clear to me what was going wrong in actual components before this fix.

File: packages/@lwc/template-compiler/src/parser/attribute.ts [hunk 1]
```diff
@@ -261,7 +261,7 @@ function shouldCamelCaseAttribute(element: IRElement, attrName: string) {
 }
 
 export function attributeToPropertyName(element: IRElement, attrName: string): string {
-    if (!shouldCamelCaseAttribute) {
+    if (!shouldCamelCaseAttribute(element, attrName)) {
         return attrName;
     }
     return toPropertyName(ATTRS_PROPS_TRANFORMS[attrName] || attrName);
```
Comment (ekashida) [inline packages/@lwc/template-compiler/src/parser/attribute.ts:267]: Is it just me, or should this be something like:
```suggestion
    return ATTRS_PROPS_TRANFORMS[attrName] || toPropertyName(attrName);
```

Comment (ravijayaramappa) [reply packages/@lwc/template-compiler/src/parser/attribute.ts]: You are right, `ATTRS_PROPS_TRANFORMS[attrName]` will always return a valid property name. No need to further transform it.

**Comment by ravijayaramappa** [comment]
@pmdartus @ekashida Spent time to debug this flow and discovered there was some redundancy in the code path. I have updated it. 

If it's too hard to reason about or you feel the change is unnecessary, please express your opinion and I will close the PR. When I started it, it was supposed to be a minor fix, but after spending time I understood the code better and thought of this simplification. 😄

File: packages/@lwc/template-compiler/src/parser/attribute.ts [hunk 2]
```diff
@@ -211,24 +211,20 @@ function isInputStateAttribute(element: IRElement, attrName: string) {
...
     }
 
-    // Handle global attrs (common to all tags) and special attribute (role, aria, key, is, data-).
-    if (isGlobalHtmlAttribute(attrName) || isAriaOrDataOrFmkAttribute(attrName)) {
-        return true;
-    }
-
     // Handle input tag value="" and checked attributes that are only used for state initialization.
     // Because .setAttribute() won't update the value, those attributes should be considered as props.
     if (isInputStateAttribute(element, attrName)) {
```
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/attribute.ts:221]: I would recommend inline the function here. The comment above explains what the `isInputStateAttribute` doing. Moving the content in place would help the overall code readability.

```suggestion
    if (element.tag === 'input' && (attrName === 'value' || attrName === 'checked')) {
```

File: packages/@lwc/template-compiler/src/parser/attribute.ts [hunk 3]
```diff
@@ -253,16 +249,11 @@ function isTemplateDirective(attrName: string): boolean {
...
-}
-
-export function attributeToPropertyName(element: IRElement, attrName: string): string {
-    if (!shouldCamelCaseAttribute) {
-        return attrName;
-    }
-    return toPropertyName(ATTRS_PROPS_TRANFORMS[attrName] || attrName);
+/**
+ * convert attribute name from kebab case to camel case property name
+ * Note: Should not invoke this function for data attribute
```
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/attribute.ts:254]: Why not?

File: packages/@lwc/template-compiler/src/parser/attribute.ts [hunk 4]
```diff
@@ -253,16 +249,11 @@ function isTemplateDirective(attrName: string): boolean {
...
-    return !isSvgTag && !isDataAttributeOrFmk;
-}
-
-export function attributeToPropertyName(element: IRElement, attrName: string): string {
-    if (!shouldCamelCaseAttribute) {
-        return attrName;
-    }
-    return toPropertyName(ATTRS_PROPS_TRANFORMS[attrName] || attrName);
+/**
+ * convert attribute name from kebab case to camel case property name
```
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/attribute.ts:253]:
```suggestion
 * Convert attribute name from kebab case to camel case property name
```

File: packages/@lwc/template-compiler/src/parser/index.ts [hunk 5]
```diff
@@ -749,6 +749,10 @@ export default function parse(source: string, state: State): TemplateParseResult
                 }
             }
 
+            // the if branch handles
+            // 1. All attributes for standard elements except 1 case are handled as attributes
+            //    * value and checked attribute for input element is treated as property
```
Comment (pmdartus) [inline packages/@lwc/template-compiler/src/parser/index.ts:754]: > * value and checked attribute for input element is treated as property

This is an internal detail of this function. I wouldn't add a comment here if we change the implementation.

```suggestion
```

**Review by ekashida — APPROVED** [review]
LGTM, although I'm still uneasy about the fact that we're fixing bugs that don't seem to have any test coverage.

File: packages/@lwc/template-compiler/src/parser/attribute.ts (hunk 3 above)
Comment (ravijayaramappa) [reply packages/@lwc/template-compiler/src/parser/attribute.ts:254]: data attributes are not processed as properties. They are always set as attributes.

Comment (ravijayaramappa) [reply packages/@lwc/template-compiler/src/parser/attribute.ts:254]: I will remove the comment. Its an internal detail of the package.

**Comment by ravijayaramappa** [comment]
> I'm still uneasy about the fact that we're fixing bugs that don't seem to have any test coverage.
@ekashida There is a good amount of test for attribute handling. The changes at this point is simplifying the logic by removing redundant checks.
https://github.com/salesforce/lwc/tree/master/packages/%40lwc/template-compiler/src/__tests__/fixtures/attributes

**Review by pmdartus — APPROVED** [review]
🚢

**Pull request merged**